Add custom styles for `ember-power-select` addon

alex-ju opened the pull request:
### :pushpin: Summary

Add custom styles for `ember-power-select` addon

### :hammer_and_wrench: Detailed description

 - Add `ember-power-select` styles (these overrides are used to align the default `ember-power-select` styles with the rest of HDS form components and have also been tested with custom implementations in `cloud-ui`)
 - To be able to manually test these in this repo, I've added `ember-power-select` as a dev dependency and created a showcase page in the dummy app.

### :camera_flash: Screenshots
Example of basic usage in the [preview link](https://hds-components-git-add-power-select-styles-hashicorp.vercel.app/overrides/power-select)

Example of custom usage in `cloud-ui` with the `hvn-select` component (local development captures)

Default
![01 hvn-select default](https://user-images.githubusercontent.com/788096/185657526-d0a62a65-0224-454f-ab7d-1df6147fc324.png)

Focus
![02 hvn-select focus](https://user-images.githubusercontent.com/788096/185657530-c0286c48-5577-4500-bdc8-f0c77471b639.png)

Expanded
![03 hvn-select open](https://user-images.githubusercontent.com/788096/185657532-5517fa40-f436-4704-9969-77067cbb493b.png)


### :link: External links

[Jira ticket for PowerSelect](https://hashicorp.atlassian.net/browse/HDS-516)
[Jira ticket for PowerSelectMultiple](https://hashicorp.atlassian.net/browse/HDS-565)

***

### 👀 How to review

<!-- Suggest how it's best for the reviewer to review the code (choose one, or remove) -->
👉 Review commit-by-commit

Reviewer's checklist:

- [ ] +1 Percy if applicable
- [ ] Confirm that PR has a changelog update via [Changesets](https://github.com/changesets/changesets) if needed

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.

MelSumner reviewed (approved):
:shipit: no objections from what I can tell. 👍

heatherlarsen reviewed (approved):
🎨 🎉 🙌

meirish reviewed (changes requested):
Is this meant to style _all_ instances of power-select? Or just the HVN picker in cloud-ui? I think most of our apps are using power-select in some manner, so this change might have a bigger blast radius than you expect.

Just want to call that out before it's merged - apologies if you've already considered that when making this change!

meirish commented on packages/components/ember-cli-build.js:
```diff
@@ -9,8 +9,12 @@ module.exports = function (defaults) {
       ],
     },
+    'ember-power-select': {
+      theme: false,
```
Do apps need to do this in their `ember-cli-build.js` too? If so, could this be added to the docs?

alex-ju replied:
As far as I can tell, this is [the default configuration](https://github.com/hashicorp/cloud-ui/blob/master/hcp/ember-cli-build.js#L147-L149), but yes it's worth mentioning that these styles assume you use `ember-power-select` with the default theme.

meirish commented:
Oh I clicked through the preview, and it looks like the styles work on the normal select too 🎉 . Do you intend to style the other modes too? with the multiselect, search, etc?

alex-ju commented:
> Is this meant to style _all_ instances of power-select? Or just the HVN picker in cloud-ui? I think most of our apps are using power-select in some manner, so this change might have a bigger blast radius than you expect.

> Just want to call that out before it's merged - apologies if you've already considered that when making this change!

> Oh I clicked through the preview, and it looks like the styles work on the normal select too 🎉 . Do you intend to style the other modes too? with the multiselect, search, etc?

It is meant to style all generic instances. I have studied the addon for a bit and tested it in 'power-select' and 'basic-dropdown' modes, and with custom markup, namely `triggerComponent`, `selectedItemComponent`, `beforeOptionsComponent` (this is where the search usually sits), `optionsComponent`, `afterOptionsComponent` and `searchMessageComponent`. I've only tested it in `cloud-ui`, but will try and spin up `nomad`, `consul` and `atlas` which contain the other instances we identified. It doesn't cover multisearch (`PowerSelectMultiple`) which is currently only found in `atlas`. I will have to get back to the team on this before committing to an answer.

didoo commented on packages/components/tests/dummy/app/controllers/components/power-select.js:
```diff
@@ -0,0 +1,9 @@
+import { action } from '@ember/object';
+
+export default class PowerSelectController extends Controller {
+  // notice: this is used as "noop" function for the onDismiss callback of the Alert component
```
the comment refers to the "Alert" component, but this is the "PowerSelect" one

didoo commented on packages/components/app/styles/components/power-select.scss:
```diff
@@ -0,0 +1,183 @@
+  font-size: 0.875rem; // 14px
+
+  .ember-basic-dropdown-trigger,
+  .ember-power-select-trigger {
+    background-color: var(--token-form-control-base-surface-color-default);
+    background-image: var(--token-form-select-background-image-data-url), url("data:image/svg+xml,%3Csvg viewBox='0 0 16 16' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath d='M14.78 4.28a.75.75 0 00-1.06-1.06l-7.97 7.97-3.47-3.47a.75.75 0 00-1.06 1.06l4 4a.75.75 0 001.06 0l8.5-8.5z' fill='%231060ff'/%3E%3C/svg%3E");
+    background-position: right var(--token-form-select-background-image-position-right-x) center, top 8px left 20px;
+    background-repeat: no-repeat;
+    background-size: var(--token-form-select-background-image-size) var(--token-form-select-background-image-size);
+    border: var(--token-form-control-border-width) solid var(--token-form-control-base-border-color-default);
+    border-radius: var(--token-form-control-border-radius);
+    box-shadow: var(--token-elevation-low-box-shadow);
+    color: var(--token-form-control-base-foreground-value-color);
+    max-width: 100%;
+    min-height: 34px;
+    padding: calc(var( --token-form-control-padding) + 1px);
+    padding-left: 44px;
+    padding-right: calc(var( --token-form-control-padding) + 24px); // extra space for the icon
```
I LOVE how we can use the component-specific design tokens for this! 😍

didoo commented on packages/components/tests/dummy/app/templates/components/power-select/after-options.hbs:
```diff
@@ -0,0 +1,3 @@
+<div class="hds-power-select_after-options">
```
Is this file needed? 🤔

alex-ju replied:
We use this to exemplify how a custom component/template block can be passed in. It's called in [one of the examples](https://github.com/hashicorp/design-system/pull/539/files/4c181a89655938a0d004db044bb82273140f0f12#diff-ac16ea0a6f51865dfe0a911f35693d355234436b89274cad9807836ab60e3a89R98).

didoo replied:
OK. I can't think of better alternatives, so let's leave it like this.

didoo commented on packages/components/tests/dummy/app/templates/components/power-select.hbs:
```diff
@@ -0,0 +1,107 @@
+{{page-title "PowerSelect Component"}}
```
There's no link to this page in the "index", is it intentional? Otherwise probably should leave under "Forms" (something like `Form / PowerSelect` maybe?).

alex-ju replied:
Was not sure where to place it at the time

didoo commented on packages/components/tests/dummy/app/templates/components/power-select.hbs:
```diff
@@ -0,0 +1,107 @@
+<h2 class="dummy-h2">PowerSelect</h2>
+
+<section>
+  <h3 class="dummy-h3" id="overview"><a href="#overview" class="dummy-link-section">§</a> Overview</h3>
```
Probably I would expand this overview, explaining that is not an HDS component but just the styling (we're essentially "re-skinning" it to make is more similar to the other form controls).

didoo commented on packages/components/tests/dummy/app/templates/components/power-select.hbs:
```diff
@@ -0,0 +1,107 @@
+</section>
+
+<section data-test-percy>
+  <h3 class="dummy-h3" id="showcase"><a href="#showcase" class="dummy-link-section">§</a> Showcase</h3>
```
Probably before the "Showcase" I would add a "How to use" section in which:
- explain that these styles assume you use ember-power-select with the default theme (see [comment here](https://github.com/hashicorp/design-system/pull/539#discussion_r951247990))
- explain how to use the specific `.hds-power-select` class (initially I thought it should be applied to the component itself, in reality it needs to be applied to a wrapping DOM node).
- explain how to use the `.hds-power-select_after-options`class (move the explanation from the showcase to the howto)

didoo commented on packages/components/tests/dummy/app/templates/components/power-select.hbs:
```diff
@@ -0,0 +1,107 @@
+  <h4 class="dummy-h4">States</h4>
+  <p class="dummy-text-small">Default</p>
+
+  {{! This below is just an example of invocation, to get started }}
```
Is this comment still relevant? seems not related to the content below.

alex-ju replied:
sure not, will remove it

didoo reviewed (commented):
@alex-ju I did a first pass, and added a few comments. Once you've decided how to "distribute" the styles (via component, as normal styles, as extra Sass to import) let me know and I'll do a final review.

One extra thing: while reading the CSS code, I thought that I'm sure in your "journey" discovering all the internals and quirks of this component you gained a lot of knowledge. Maybe have a read at the CSS file with the styles and add comments where you think it would be helpful for someone in the future to know certain things :) (included your future self)

alex-ju commented:
> Oh I clicked through the preview, and it looks like the styles work on the normal select too 🎉 . Do you intend to style the other modes too? with the multiselect, search, etc?

@meirish an update on this: we found a few instances of `PowerSelectMultiple` only in atlas and terraform registry and having discussed this with the team we will provide custom styles for these as well.

alex-ju commented:
This PR is good for another review.

@didoo I addressed your comments in this PR in conjunction with the [decisions agreed during the sync](https://docs.google.com/document/d/1xXxMv70KCOEmAe_l8mfzrnbQ3EGrw3v6Qt107RPyWuA/edit#bookmark=id.twbdrk2s4sxl).

In short in the additional commits I've:
 - moved the PowerSelect styles to a new 'Overrides' section (hopefully we won't have many of these) to reflect it's bespoke nature
 - added the 'Overview' and 'How to use' sections trying to better document these overrides
 - applied the stylelint rules with an exception on class names pattern (as we're not in control of these names)

@heatherlarsen I have updated this PR to style the PowerSelectMultiple based on your designs. There is one limitation: the way the tags are being marked-up by the vendor doesn't allow us to use the hover we have in our Tag component – otherwise I did my best to match the design, but let me know if there's something I missed or you'd want me to adjust.

didoo commented on packages/components/app/styles/@hashicorp/power-select-overrides.scss:
```diff
@@ -0,0 +1,225 @@
+// and are meant to provide overrides through specificity
+//
+
+/* stylelint-disable selector-class-pattern */
```
👍

didoo commented on packages/components/tests/dummy/app/templates/overrides/power-select.hbs:
```diff
@@ -0,0 +1,284 @@
+  <CodeBlock
+    @code='
+      @import "@hashicorp/design-system-components";
+      @import "@hashicorp/power-select-overrides";
```
_maybe_ I would be over-specific and explicit, and call the file `@hashicorp/design-system-power-select-overrides`

alex-ju replied:
agree, better over-specific than ambiguous – will update the file name (and references to it)

didoo commented on packages/components/tests/dummy/app/templates/index.hbs:
```diff
@@ -175,6 +175,16 @@
           </LinkTo>
         </li>
       </ol>
+      <h2 class="dummy-h2">
+        Overrides
+      </h2>
+      <ol>
+        <li class="dummy-paragraph">
+          <LinkTo @route="overrides.power-select">
+            PowerSelect
+          </LinkTo>
+        </li>
+      </ol>
```
I like it!

didoo reviewed (approved):
all perfect! 🙌

(left only a couple of minor comments, non blocking)

heatherlarsen commented:
> the way the tags are being marked-up by the vendor doesn't allow us to use the hover we have in our Tag component

That's ok. These look great. Thanks Alex 👍

KristinLBradley commented on packages/components/tests/dummy/app/templates/overrides/power-select.hbs:
```diff
@@ -0,0 +1,284 @@
+  <CodeBlock
+    @language="markup"
+    @code='
+    <div class="hds-power-select">
```
Question: Is it not possible/desirable to add a class name to the "PowerSelect" component itself vs. wrapping it with an extra div?

alex-ju replied:
Great question, @KristinLBradley! I would've loved that to be possible, but applying a `class="hds-power-select"` attribute to a `<PowerSelect>` instance results in something like this:
```
<div class="ember-basic-dropdown">
  <div class="ember-view ember-basic-dropdown-trigger hds-power-select">...</div>
  <div class="ember-basic-dropdown-content">...</div>
</div>
```
So the class is not being applied to the outermost element, but to the dropdown trigger, which doesn't help with styling the dropdown content.

alex-ju replied:
We could use adjacent sibling selector, but we needed some adjustments to the element that holds the `ember-basic-dropdown` hence I decided to use that additional wrapper element, which you can imagine I'm not fond of.

KristinLBradley replied:
I see. Thanks for explaining.
